### PR TITLE
MLDB-2109 serial proc progress enh

### PR DESCRIPTION
**Pull request by FinchPowers**

### CODE REVIEW

File: core/procedure.cc [hunk 1]
```diff
@@ -317,8 +319,32 @@ run(const ProcedureRunConfig & run,
...
+        return onProgress(progress);
+    };
+
+    for (int i = 0; i < steps.size(); ++i ) {
+        auto & s = steps[i];
+        iterationStep->value = (float)i / steps.size() * 100;
+
+        bool keepGoing = onProg(Json::Value{});
+        if (!keepGoing) {
+                throw MLDB::CancellationException("Procedure mock cancelled");
```
Comment (guyd) [inline core/procedure.cc]: Error message should be changed.

File: core/procedure.cc [hunk 2]
```diff
@@ -317,8 +319,32 @@ run(const ProcedureRunConfig & run,
...
+            progress["subProgress"] = data;
+        }
+        return onProgress(progress);
+    };
+
+    for (int i = 0; i < steps.size(); ++i ) {
+        auto & s = steps[i];
+        iterationStep->value = (float)i / steps.size() * 100;
+
+        bool keepGoing = onProg(Json::Value{});
```
Comment (guyd) [inline core/procedure.cc]: Fishy.  Usually, the progress payload should cumulate the progress information from all steps.

File: testing/MLDB-621_serial_procedure_deadlock.py [hunk 3]
```diff
@@ -1,16 +0,0 @@
-#
-# MLDB-621_serial_procedure_deadlock.py
```
Comment (guyd) [inline testing/MLDB-621_serial_procedure_deadlock.py]: Why removing it?  Isn't still detecting that recursive references aren't deadlocking?

File: testing/serial_procedure_test.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
+#
+# serial_procedure_test.py
+# datacratic, 2015
```
Comment (guyd) [inline testing/serial_procedure_test.py]: 2015->2017

**Comment by guyd** [comment]
+1 once the concerns have been reviewed

File: testing/serial_procedure_test.py (hunk 4 above)
Comment (FinchPowers) [inline testing/serial_procedure_test.py]: It's the original creation of the original file. I'm unsure whether I should bump it or not.

File: core/procedure.cc (hunk 2 above)
Comment (FinchPowers) [inline core/procedure.cc]: How do we do that? The problem is that the "progress" is lost whenever a procedure finishes.

Comment (guyd) [inline core/procedure.cc]: Above serial progress should contain a progress step of each serial step (e.g. using the serial step id).  Pass the index `i` as part of onProg param (you will need to bind it before calling `run` on the serial step).  Whenever, `i` is greater than the last value, increment the serial progress with `next`.

File: core/procedure.cc [hunk 5]
```diff
@@ -317,10 +319,37 @@ run(const ProcedureRunConfig & run,
...
+        Json::Value progress = jsonEncode(serialProgress);
+        if (!data.empty()) {
+            progress["subProgress"] = data;
+        }
+        return onProgress(progress);
+    };
+
+    for (int i = 0; i < steps.size(); ++i ) {
+        auto & s = steps[i];
+        //iterationStep->value = (float)i / steps.size() * 100;
```
Comment (guyd) [inline core/procedure.cc]: Commented code.

**Pull request merged**